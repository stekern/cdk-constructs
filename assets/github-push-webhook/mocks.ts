import { ForwardingRule, DbPushEvent } from "./types"

export const pushEvent: DbPushEvent = {
  PK: "pk",
  SK: "sk",
  schemaVersion: "0.1",
  branch: "dev",
  isDefaultBranch: true,
  payload: {
    ref: "refs/heads/dev",
    before: "83054c3ad04a03ed653f63e9ff2adb47c02c097d",
    after: "5b48e53fe13c778e6fe081097dff053215a42045",
    repository: {
      id: 23456789,
      node_id: "A_bcdefgh",
      name: "my-repo",
      full_name: "ghuser/my-repo",
      private: false,
      owner: {
        name: "ghuser",
        email: "ghuser@users.noreply.github.com",
        login: "ghuser",
        id: 12345678,
        node_id: "B_cdefghi",
        avatar_url: "https://avatars.githubusercontent.com/u/12345678?v=4",
        gravatar_id: "",
        url: "https://api.github.com/users/ghuser",
        html_url: "https://github.com/ghuser",
        followers_url: "https://api.github.com/users/ghuser/followers",
        following_url:
          "https://api.github.com/users/ghuser/following{/other_user}",
        gists_url: "https://api.github.com/users/ghuser/gists{/gist_id}",
        starred_url:
          "https://api.github.com/users/ghuser/starred{/owner}{/repo}",
        subscriptions_url: "https://api.github.com/users/ghuser/subscriptions",
        organizations_url: "https://api.github.com/users/ghuser/orgs",
        repos_url: "https://api.github.com/users/ghuser/repos",
        events_url: "https://api.github.com/users/ghuser/events{/privacy}",
        received_events_url:
          "https://api.github.com/users/ghuser/received_events",
        type: "User",
        site_admin: false,
      },
      html_url: "https://github.com/ghuser/my-repo",
      description: "Experimental AWS CDK construct library",
      fork: false,
      url: "https://github.com/ghuser/my-repo",
      forks_url: "https://api.github.com/repos/ghuser/my-repo/forks",
      keys_url: "https://api.github.com/repos/ghuser/my-repo/keys{/key_id}",
      collaborators_url:
        "https://api.github.com/repos/ghuser/my-repo/collaborators{/collaborator}",
      teams_url: "https://api.github.com/repos/ghuser/my-repo/teams",
      hooks_url: "https://api.github.com/repos/ghuser/my-repo/hooks",
      issue_events_url:
        "https://api.github.com/repos/ghuser/my-repo/issues/events{/number}",
      events_url: "https://api.github.com/repos/ghuser/my-repo/events",
      assignees_url:
        "https://api.github.com/repos/ghuser/my-repo/assignees{/user}",
      branches_url:
        "https://api.github.com/repos/ghuser/my-repo/branches{/branch}",
      tags_url: "https://api.github.com/repos/ghuser/my-repo/tags",
      blobs_url: "https://api.github.com/repos/ghuser/my-repo/git/blobs{/sha}",
      git_tags_url:
        "https://api.github.com/repos/ghuser/my-repo/git/tags{/sha}",
      git_refs_url:
        "https://api.github.com/repos/ghuser/my-repo/git/refs{/sha}",
      trees_url: "https://api.github.com/repos/ghuser/my-repo/git/trees{/sha}",
      statuses_url:
        "https://api.github.com/repos/ghuser/my-repo/statuses/{sha}",
      languages_url: "https://api.github.com/repos/ghuser/my-repo/languages",
      stargazers_url: "https://api.github.com/repos/ghuser/my-repo/stargazers",
      contributors_url:
        "https://api.github.com/repos/ghuser/my-repo/contributors",
      subscribers_url:
        "https://api.github.com/repos/ghuser/my-repo/subscribers",
      subscription_url:
        "https://api.github.com/repos/ghuser/my-repo/subscription",
      commits_url: "https://api.github.com/repos/ghuser/my-repo/commits{/sha}",
      git_commits_url:
        "https://api.github.com/repos/ghuser/my-repo/git/commits{/sha}",
      comments_url:
        "https://api.github.com/repos/ghuser/my-repo/comments{/number}",
      issue_comment_url:
        "https://api.github.com/repos/ghuser/my-repo/issues/comments{/number}",
      contents_url:
        "https://api.github.com/repos/ghuser/my-repo/contents/{+path}",
      compare_url:
        "https://api.github.com/repos/ghuser/my-repo/compare/{base}...{head}",
      merges_url: "https://api.github.com/repos/ghuser/my-repo/merges",
      archive_url:
        "https://api.github.com/repos/ghuser/my-repo/{archive_format}{/ref}",
      downloads_url: "https://api.github.com/repos/ghuser/my-repo/downloads",
      issues_url: "https://api.github.com/repos/ghuser/my-repo/issues{/number}",
      pulls_url: "https://api.github.com/repos/ghuser/my-repo/pulls{/number}",
      milestones_url:
        "https://api.github.com/repos/ghuser/my-repo/milestones{/number}",
      notifications_url:
        "https://api.github.com/repos/ghuser/my-repo/notifications{?since,all,participating}",
      labels_url: "https://api.github.com/repos/ghuser/my-repo/labels{/name}",
      releases_url: "https://api.github.com/repos/ghuser/my-repo/releases{/id}",
      deployments_url:
        "https://api.github.com/repos/ghuser/my-repo/deployments",
      created_at: 1661001460,
      updated_at: "2022-08-22T20:21:45Z",
      pushed_at: 1668277802,
      git_url: "git://github.com/ghuser/my-repo.git",
      ssh_url: "git@github.com:ghuser/my-repo.git",
      clone_url: "https://github.com/ghuser/my-repo.git",
      svn_url: "https://github.com/ghuser/my-repo",
      homepage: "",
      size: 159,
      stargazers_count: 0,
      watchers_count: 0,
      language: "TypeScript",
      has_issues: true,
      has_projects: true,
      has_downloads: true,
      has_wiki: true,
      has_pages: false,
      forks_count: 0,
      mirror_url: null,
      archived: false,
      disabled: false,
      open_issues_count: 1,
      license: null,
      allow_forking: true,
      is_template: false,
      web_commit_signoff_required: false,
      topics: ["aws", "cdk"],
      visibility: "public",
      forks: 0,
      open_issues: 1,
      watchers: 0,
      default_branch: "dev",
      stargazers: 0,
      master_branch: "dev",
    },
    pusher: {
      name: "ghuser",
      email: "ghuser@users.noreply.github.com",
    },
    sender: {
      login: "ghuser",
      id: 12345678,
      node_id: "B_cdefghi",
      avatar_url: "https://avatars.githubusercontent.com/u/12345678?v=4",
      gravatar_id: "",
      url: "https://api.github.com/users/ghuser",
      html_url: "https://github.com/ghuser",
      followers_url: "https://api.github.com/users/ghuser/followers",
      following_url:
        "https://api.github.com/users/ghuser/following{/other_user}",
      gists_url: "https://api.github.com/users/ghuser/gists{/gist_id}",
      starred_url: "https://api.github.com/users/ghuser/starred{/owner}{/repo}",
      subscriptions_url: "https://api.github.com/users/ghuser/subscriptions",
      organizations_url: "https://api.github.com/users/ghuser/orgs",
      repos_url: "https://api.github.com/users/ghuser/repos",
      events_url: "https://api.github.com/users/ghuser/events{/privacy}",
      received_events_url:
        "https://api.github.com/users/ghuser/received_events",
      type: "User",
      site_admin: false,
    },
    installation: {
      id: 34567890,
      node_id: "C_defghij",
    },
    created: false,
    deleted: false,
    forced: false,
    base_ref: null,
    compare:
      "https://github.com/ghuser/my-repo/compare/83054c3ad04a...5b48e53fe13c",
    commits: [
      {
        id: "5b48e53fe13c778e6fe081097dff053215a42045",
        tree_id: "914f3a17f6c1c2fbcef2671fb80fb4e14c4c029b",
        distinct: true,
        message: "fix: potential fix for bsd-specific behavior for cp utility",
        timestamp: "2022-11-12T19:29:58+01:00",
        url: "https://github.com/ghuser/my-repo/commit/5b48e53fe13c778e6fe081097dff053215a42045",
        author: {
          name: "ghuser",
          email: "ghuser@users.noreply.github.com",
          username: "ghuser",
        },
        committer: {
          name: "ghuser",
          email: "ghuser@users.noreply.github.com",
          username: "ghuser",
        },
        added: [],
        removed: [],
        modified: ["package.json"],
      },
    ],
    head_commit: {
      id: "5b48e53fe13c778e6fe081097dff053215a42045",
      tree_id: "914f3a17f6c1c2fbcef2671fb80fb4e14c4c029b",
      distinct: true,
      message: "fix: potential fix for bsd-specific behavior for cp utility",
      timestamp: "2022-11-12T19:29:58+01:00",
      url: "https://github.com/ghuser/my-repo/commit/5b48e53fe13c778e6fe081097dff053215a42045",
      author: {
        name: "ghuser",
        email: "ghuser@users.noreply.github.com",
        username: "ghuser",
      },
      committer: {
        name: "ghuser",
        email: "ghuser@users.noreply.github.com",
        username: "ghuser",
      },
      added: [],
      removed: [],
      modified: ["package.json"],
    },
  },
}
export const rule: ForwardingRule = {
  owner: "ghuser",
  repo: "my-repo",
  channel: "#my-channel",
}
