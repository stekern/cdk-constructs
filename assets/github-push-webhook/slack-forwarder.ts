import { DbPushEvent, ForwardingRule } from "./types"
import { URL } from "url"
import * as https from "https"
import { DynamoDBStreamEvent } from "aws-lambda"
import DynamoDB from "aws-sdk/clients/dynamodb"

type SlackPayload = {
  channel?: string
  icon_emoji: string
  username: string
  text: string
  blocks: {
    type: string
    text?: {
      type: string
      text: string
    }
    fields?: {
      type: string
      text: string
    }[]
    elements?: {
      type: string
      text: string
    }[]
  }[]
  attachments?: {
    footer?: string
    text: string
    mrkdwn_in?: string[]
  }[]
}

const httpRequest = (
  params: https.RequestOptions,
  payload?: string,
  parseJson?: boolean,
): Promise<Record<string, unknown> | string | undefined> => {
  return new Promise(function (resolve, reject) {
    const req = https.request(params, function (res) {
      if (res.statusCode! < 200 || res.statusCode! >= 300) {
        return reject(
          new Error(`Received non-200 status code ${res.statusCode!}`),
        )
      }
      let body = ""
      res.on("data", function (chunk) {
        body += chunk
      })
      res.on("end", function () {
        let result
        if (parseJson) {
          try {
            result = JSON.parse(body) as Record<string, unknown>
          } catch (e) {
            reject(new Error("Failed to deserialize response as JSON"))
          }
        }
        resolve(result)
      })
    })
    req.on("error", function (err) {
      reject(err)
    })
    if (payload) {
      req.write(payload)
    }
    req.end()
  })
}

const createSlackPayload = (
  rule: ForwardingRule,
  pushEvent: DbPushEvent,
): SlackPayload => {
  return {
    channel: rule.channel,
    icon_emoji: ":twisted_rightwards_arrows:",
    username: "GitHub Integration",
    text: `Commit(s) pushed to repository ${pushEvent.payload.repository.full_name}`,
    blocks: [
      {
        type: "section",
        text: {
          type: "mrkdwn",
          text: `<${pushEvent.payload.compare}|${
            pushEvent.payload.commits.length
          } new commit${
            pushEvent.payload.commits.length > 1 ? "s" : ""
          }> pushed to <${pushEvent.payload.repository.html_url}/tree/${
            pushEvent.branch
          }|\`${pushEvent.branch}\`> by <${
            pushEvent.payload.sender.html_url
          }|\`${pushEvent.payload.sender.login}\`>`,
        },
      },
    ],
    attachments: [
      {
        footer: `<${pushEvent.payload.repository.html_url}|${pushEvent.payload.repository.full_name}>`,
        mrkdwn_in: ["text"],
        text: pushEvent.payload.commits
          .map(
            (commit) =>
              `<${commit.url}|\`${commit.id.substring(0, 8)}\`> - ${
                // Split at newline to avoid showing potentially long commit description
                commit.message.split(/\n/)[0]
              }`,
          )
          .join("\n"),
      },
    ],
  }
}

export const handler = async (event: DynamoDBStreamEvent) => {
  console.log(JSON.stringify(event, null, 2))
  const slackWebhookUrl = process.env.SLACK_WEBHOOK_URL
    ? new URL(process.env.SLACK_WEBHOOK_URL)
    : undefined
  const forwardingRules = (
    process.env.FORWARDING_RULES ? JSON.parse(process.env.FORWARDING_RULES) : []
  ) as ForwardingRule[]

  if (!slackWebhookUrl) {
    throw Error("Missing required environment variable")
  }
  if (forwardingRules.length === 0) {
    console.log("No Slack forwarding rules set up")
  }
  const pushEvents = event.Records.filter((r) => r.dynamodb?.NewImage).map(
    (r) => DynamoDB.Converter.unmarshall(r.dynamodb!.NewImage!) as DbPushEvent,
  )
  const payloads = forwardingRules.flatMap((forwardingRule) => {
    const matchingPushEvents = pushEvents.filter(
      (e) =>
        e.payload.repository.full_name ===
        `${forwardingRule.owner}/${forwardingRule.repo}`,
    )
    return matchingPushEvents.map((pushEvent) =>
      createSlackPayload(forwardingRule, pushEvent),
    )
  })
  const options: https.RequestOptions = {
    hostname: slackWebhookUrl.hostname,
    port: slackWebhookUrl.port,
    path: slackWebhookUrl.pathname,
    method: "POST",
    headers: {
      "Content-Type": "application/json; charset=utf-8",
    },
  }
  await Promise.all(
    payloads.map((p) => httpRequest(options, JSON.stringify(p))),
  )
}
