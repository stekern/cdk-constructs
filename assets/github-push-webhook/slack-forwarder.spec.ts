import { expect, it } from "vitest"
import { createSlackPayload } from "./slack-forwarder"
import { pushEvent, rule } from "./mocks"

it("slack payload snapshot", () => {
  const payload = createSlackPayload(rule, pushEvent)
  expect(payload).toMatchSnapshot()
})
