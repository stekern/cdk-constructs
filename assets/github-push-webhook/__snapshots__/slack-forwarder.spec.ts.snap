// Vitest Snapshot v1

exports[`payload 1`] = `
{
  "attachments": [
    {
      "footer": "<https://github.com/ghuser/my-repo|ghuser/my-repo>",
      "mrkdwn_in": [
        "text",
      ],
      "text": "<https://github.com/ghuser/my-repo/commit/5b48e53fe13c778e6fe081097dff053215a42045|\`5b48e53f\`> - fix: potential fix for bsd-specific behavior for cp utility",
    },
  ],
  "blocks": [
    {
      "text": {
        "text": "<https://github.com/ghuser/my-repo/compare/83054c3ad04a...5b48e53fe13c|1 new commit> pushed to <https://github.com/ghuser/my-repo/tree/dev|\`dev\`> by <https://github.com/ghuser|\`ghuser\`>",
        "type": "mrkdwn",
      },
      "type": "section",
    },
  ],
  "channel": "#my-channel",
  "icon_emoji": ":twisted_rightwards_arrows:",
  "text": "Commit(s) pushed to repository ghuser/my-repo",
  "username": "GitHub Integration",
}
`;

exports[`slack payload snapshot 1`] = `
{
  "attachments": [
    {
      "footer": "<https://github.com/ghuser/my-repo|ghuser/my-repo>",
      "mrkdwn_in": [
        "text",
      ],
      "text": "<https://github.com/ghuser/my-repo/commit/5b48e53fe13c778e6fe081097dff053215a42045|\`5b48e53f\`> - fix: potential fix for bsd-specific behavior for cp utility",
    },
  ],
  "blocks": [
    {
      "text": {
        "text": "<https://github.com/ghuser/my-repo/compare/83054c3ad04a...5b48e53fe13c|1 new commit> pushed to <https://github.com/ghuser/my-repo/tree/dev|\`dev\`> by <https://github.com/ghuser|\`ghuser\`>",
        "type": "mrkdwn",
      },
      "type": "section",
    },
  ],
  "channel": "#my-channel",
  "icon_emoji": ":twisted_rightwards_arrows:",
  "text": "Commit(s) pushed to repository ghuser/my-repo",
  "username": "GitHub Integration",
}
`;
